Toplevel reimplementation with leaf answer callbacks

bakaq opened the pull request:
I went and reimplemented a good part of the toplevel on top of a predicate I called `run_query_goal/4`. It passes all the current tests, and works the same as the old implementation in my manual testing. The idea is to eventually make this same predicate drive the library interface (when I figure out how to call arbitrary Prolog from the Rust side[^1]), other (pseudo) toplevels, and maybe even some other kinds of things, it's very general. It takes 4 arguments: the query term, a list of variable names (in the form given by the `variable_names(-VarNames)` option of [`read_term/3`](https://www.scryer.pl/builtins.html#read_term/3)), a callback goal expecting 3 arguments and a list of options. This predicate then runs the query and calls the callback with the leaf answer and additional information, and the callback then signals if the query should continue or stop. The possibilities for the first argument (the leaf answer) are:

- `final(false)`
- `final(exception(Exception))`
- `final(true)`
- `final(leaf_answer(Bindings, ResidualGoals, VarNames))`
- `pending(true)`
- `pending(leaf_answer(Bindings, ResidualGoals, VarNames))`

A principal functor of `final/1` means that this is the last leaf answer in the complete answer, and a principal functor of `pending/1` means that there are more leaf answers. Implementing this made me realize that this pending/final "attribute" of leaf answers is actually very important (it's not quite an iterator, in the Rust sense) and gives me ideas of how to improve the library interface further[^2]. The second argument to the callback is a list with additional information that we can use to implement extensions to this in the future (like reporting inference counts). The third argument needs to be instantiated by the callback to either `continue` or `stop` to signal if the query should stop.

Notice that this can fully represent residual goals (!), with projections (!!) and also reifies exceptions (!!!). The only improvement on this I can think of is to map the query variable names onto the exception like in the `leaf_answer/3` case, but it seems that variables lose their identity when going through a throw-catch, so I don't know if that is even possible.

I want to further break the toplevel into configurable/reusable pieces in the future, because that will allow a lot of really cool stuff, especially in the [Scryer Playground](https://play.scryer.pl/)!

Ok, now I gotta sleep, I've been coding this non-stop for 14 hours straight.

[^1]: I have some ideas! If it works we could move even more of the engine into Prolog very easily.
[^2]: It also was the first time I've felt a use for a soft cut, which was interesting to stumble on, ~and the first time I've actually used `subsumes_term/2`, very nice~ (this was actually not a place where `subsumes_term/2` helped much).

triska commented on src/toplevel.pl:
```diff
@@ -200,9 +203,144 @@
+    ;   handle_first_answer,
+        increment_answer_count,
+        show_leaf_answer(LeafAnswer, []),
+        write('.'), nl
```
No need to quote the atom&nbsp;`.` here.

bakaq replied:
I knew you would point this out. I think quoting it is clearer, as it's a character that is also used for special things (ending clauses) so it can get a bit confusing. A bit related: why does `write_canonical/1` print lists with the quoted dot?

`´´prolog
?- write_canonical([1,2,3]), nl.
'.'(1,'.'(2,'.'(3,[])))
   true.
?- X = .(1,.(2,.(3,[]))).
   % This also works.
   X = [1,2,3].
```

This seems to just make serializing lists (and especially strings of chars) even more inefficient. Is it for readability? If it is, that's exactly my point, but I don't think this should be an big concern for `write_canonical/1` in specific.

jjtolton replied:
(my guess is it's the ISO standard). :popcorn:

UWN replied:
[7.10.5 d](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/draft-7.10.5#d) requires the quoting of a single dot atom with an effective write option `quoted(true)` which means that [`writeq/1`](https://www.complang.tuwien.ac.at/ulrich/iso-prolog/conformity_testing#32) as well as `write_canonical/1` have to quote it in any context. Note that a single dot is very brittle otherwise. Add a space at the end and the dot-atom becomes an end-token.

There [are](https://stackoverflow.com/q/59821549/772868) faster and more space efficient ways of writing. But as long as #1714 and #16 are open, it does not make sense to make things faster here. For conformity, this very verbose format has to be used. Yes at least 7n+2 compared to at least n+2 with double quoted lists.  It could be that also reading could be more efficient here, [but this is not that sure](https://stackoverflow.com/q/70238387/772868).

triska commented on src/toplevel.pl:
```diff
@@ -200,9 +203,144 @@
+    read_term_from_chars(Query, QueryTerm, [variable_names(VarNames)]),
+    run_query_term(QueryTerm, VarNames, Callback_1).
+
+run_query_term(QueryTerm, VarNames, Callback_1) :-
```
A `QueryTerm` is a `Goal`?

triska commented on src/toplevel.pl:
```diff
@@ -200,9 +203,144 @@
+        write('.'), nl
+    ).
+
+show_leaf_answer(true, _) :- write(true).
```
`show` is `write`?

triska commented:
> variables lose their identity when going through a throw-catch

You can throw, as exception, a term of the form `T-Vs`, where `Vs` are the variables in&nbsp;`T`. This will yield a copy of `T` and&nbsp;`Vs`. If you have the original variables `Vs0` somewhere, then unifying `Vs` with `Vs0` will reinstate the original variables in&nbsp;`T`.

triska commented on src/toplevel.pl:
```diff
@@ -200,9 +203,144 @@
+    show_leaf_answer(LeafAnswer, []),
+    read_input2(LeafAnswer).
+toplevel_query_callback(final(LeafAnswer)) :-
+    (   subsumes_term(exception(_), LeafAnswer) ->
```
`(  LeafAnswer = exception(Exception) -> ... ` , or even better, delegate the case distinction to argument indexing?

triska commented:
Wow, thank you a lot for working on this! I added a few comments directly in the commit, and I hope you find them useful!

UWN commented:
What about a seamless integration of inference limits and/or timeouts?

jjtolton commented:
This is amazing work!  Once this is done, would you have any interest in doing an informal mini-talk or discussion about your experience reimplementing the toplevel?  It's a little beyond my current aptitude but I'm really fascinated by it because no other language I've worked with manipulates its own internals like this.  It's like a puppet working its own ($$\tiny{compact null terminated UTF-8}$$) strings!  I'd be really interest in your thoughts on it.

bakaq commented:
> What about a seamless integration of inference limits and/or timeouts?

I assume you mean this as options to `run_query_term/3` and not necessarily the toplevel. I thought about that, and I think I should at least leave space for that to be implemented in the future. If I find out how to do it I may even implement this now.

> Once this is done, would you have any interest in doing an informal mini-talk or discussion about your experience reimplementing the toplevel?

Definitely. I already worked a [whole bunch](https://github.com/mthom/scryer-prolog/pulls?q=is%3Apr+author%3Abakaq) on [low level](https://github.com/bakaq/constrained.pl) Prolog [^1] with the intent to make high level monotonic Prolog easier and more general to use. It's fascinating and I would really like to share some of the insights I learned along the way.

[^1]: Low level Prolog (implementing constraint solvers and systems to work with side effects) is a good example on where you either can't or don't want monotonic code, though bare `!/0` is still very situational even at this level and it's better to use some more structured non-monotonic abstractions (`(->)/2`, `(\+)/1`, `once/1`, `nonvar/1`, etc...).

bakaq commented:
I renamed some things, removed the old toplevel, expanded the interface with an `Options` argument and an extra argument to the callback to allow future extensions and documented `run_query/3` and `run_query_term/4`. I'm thinking of maybe pulling `exception(_)` out of `final(_)`, because that may be easier to use in practice (exception is almost always error handling).

> You can throw, as exception, a term of the form T-Vs, where Vs are the variables in T.

This sounds interesting, but what I wanted here is to be able to trace variable identity for _any_ exception[^1], to have things like this:

```prolog
?- throw(a(A, B)).
   % Current behavior
   throw(a(_41,_42)).
?- throw(a(A, B)).
   % Nice behavior I wanted to implement
   throw(a(A, B)).
```

Is there a fundamental reason why variables need to be copied and lose their identity in exceptions, or why you wouldn't want the "nice behavior" here? Related to this, and I already talked about this before, currently there is an inconsistency between printing of exceptions that are errors (in the sense of having principal functor `error/2` and exceptions that aren't:

```prolog
?- throw(a(A, B)).
   throw(a(_41,_42)).
?- throw(error(A, B)).
   % Strips away the throw/1
   error(_41,_42).
``` 

I guess this is for readability, but I think this is maybe the last detail that makes the toplevel not completely follow the philosophy of "the complete answer is logically equivalent to the query"[^2]. If readability is the only argument here, I don't think it matters much. Let's just implement basic comment error pretty printing and readability is solved. In fact I think that I may work on that next, doesn't seem that difficult, although it won't be that useful yet because errors still don't have much information. The vision:

```prolog
?- throw(a(A, B)).
   throw(a(A,B)).
?- throw(error(instantiation_error,example/2).
   % ERROR: instantiation_error
   % - culprit: example/2
   throw(error(instantiation_error,culprit/2).
```

[^1]: And also doing an unification may trigger some constraint propagation and have some crazy side effects which I think we should avoid as much as possible. But if a constraint implementation does that for unifying two variables then I think it's probably the constraint fault.
[^2]: Although `error/2` is actually a predicate, so it still _kinda_ works.

bakaq commented:
I added a mechanism to stop the query with another argument to the callback. This is needed for when you press enter in the toplevel and also in the `Drop` implementation of `QueryState`. Previously I did a hack using `throw/1` and `catch/3` with an special atom for control flow, which just seemed wrong.

triska commented:
> what I wanted here is to be able to trace variable identity for any exception

You can use `term_variables/2` to obtain all variables that occur in the query.

triska commented:
This is awesome, thank you a lot for all this work! 

In the description of the PR, you mention "a use for a soft cut"; which part of your code would benefit from this? Scryer currently does not provide a soft&nbsp;cut; I see in your code you are able to inspect the WAM to deduce interesting information about choice&nbsp;points, maybe this part can be separated as a construct, at least to improve readability of the current implementation.

mthom commented:
@bakaq I'll try this out soon and tell if I have any notes. thanks for your patience.

mthom commented:
@bakaq very nice! there's some trailing whitespace in toplevel.pl but it all looks good

UWN commented on src/toplevel.pl:
```diff
@@ -170,8 +170,8 @@
 %% Enable op declarations with lists of operands, i.e.,
 %% :- op(900, fy, [$,@]).
 
-user:term_expansion((:- op(Pred, Spec, Ops)), OpResults) :- 
```
That `Pred` should be `Pri` for priority.

bakaq replied:
Well, that was there before this PR, but I guess I could fix it too.

bakaq commented:
Is there anything else to be done here? I'm currently trying to use this to implement `Machine::run_query()` in #2582 with support for residual goals and such.

mthom commented:
no, was just waiting for the correction in reply to @UWN's comment. merging now
